tests/difference/core/quint_model/ccv.qnt

```diff
@@ -0,0 +1,877 @@
+
+    // The maximum time duration between sending any VSCPacket to any consumer chain and receiving the
+    // corresponding VSCMaturedPacket, without timing out the consumer chain and consequently removing it.
+    const VscTimeout: int
+
+    // The timeoutTimestamp for sent packets. Can differ by chain.
+    const CcvTimeout: Chain -> int
+
+    // ===================
+    // PROTOCOL LOGIC contains the meat of the protocol
+    // functions here roughly correspond to API calls that can be triggered from external sources
+    // ===================
+    
+    // the power of a validator on the provider chain is changed to the given amount. We do not care how this happens,
+    // e.g. via undelegations, or delegations, ...
+    pure def votingPowerChange(currentState: ProtocolState, validator: Node, amount: int): Result = {
+            if (amount < 0) {
+                Err("Voting power changes must be positive")
+            } else {
+                pure val currentValidatorSet = currentState.providerState.chainState.currentValidatorSet
+                pure val newValidatorSet = getUpdatedValidatorSet(currentValidatorSet, validator, amount)
+                // set the validator set changed flag
+                val newProviderState = currentState.providerState.with(
+                    "providerValidatorSetChangedInThisBlock", true
+                )
+                pure val tmpState = currentState.with(
+                    "providerState", newProviderState
+                )
+                pure val newState = setProviderValidatorSet(tmpState, newValidatorSet)
+                Ok(newState)
+            }
+    }
+
+    // Delivers the next queued VSCMaturedPacket from a consumer chain to the provider chain.
+    // Only argument is the consumer chain, from which the packet will be delivered.
+    // If this packet will time out on the provider on delivery,
+    // the consumer will be dropped.
+    // The first return is the result of the operation, the second result is a boolean
+    // that indicates whether the consumer timed out or not.
+    // If the result has an error, the second return should be ignored.
+    pure def deliverPacketToProvider(currentState: ProtocolState, sender: Chain): (Result, bool) = {
+        if (not(isCurrentlyConsumer(sender, currentState.providerState))) {
+            (Err("Sender is not currently a consumer - must have 'running' status!"), false)
+        } else if (length(currentState.consumerStates.get(sender).outstandingPacketsToProvider) == 0) {
+            (Err("No outstanding packets to deliver"), false)
+        } else {
+            val packet = currentState.consumerStates.get(sender).outstandingPacketsToProvider.head()
+            // if the packet has timed out, drop the consumer. its state doesn't matter anymore
+            val timeout = CcvTimeout.get(sender)
+            if(packet.sendingTime + CcvTimeout.get(sender) < currentState.providerState.chainState.lastTimestamp) {
```

Comment:
If I understand this correctly, the `packet.sendingTime` corresponds to time set by the consumer chain, but here we compare against the provider's time. I guess this is done for simplicity? Because otherwise, a provider `p` would need to store the latest time `p` received a packet from a consumer and then go through all those latest times to timeout consumer chains?

Reply:
I think this is correct and not a simplification -
the provider receives a packet.
packet.sendingTime + CcvTimeout.get(sender) is at what timestamp on the receiving chain the packet will time out.

Notice this is *not* the VscTimeout - the CcvTimeout is related to the timeout timestamp written into the ibc packets.

If you disagree or this doesn't make sense, let me know and let's have a sync

Reply:
Yes, that makes sense. Thanks for the explanation. (I'm resolving this).

Reply:
No, it should be the timestamp on the consumer.

Here is the explanation in more detail:
The consumer chain sends the packet at packet.sendingTime.
In the real code, IBC packets store the time *on the destination chain* at which they are meant to time out.
This is typically the sending time *on the sender* + a certain delay (the CcvTimeout in our case).
In our case, the timeout time is therefore packet.sendingTime + CcvTimeout.get(sender).
Remember, this is the timestamp *on the destination chain* at which the packet times out.
So we need to compare packet.sendingTime + CcvTimeout.get(sender) with the timestamp on the destination chain (so provider.time, or currentState.providerState.chainState.lastTimestamp, in our case).

References: https://github.com/cosmos/ibc/tree/main/spec/core/ics-004-channel-and-packet-semantics#definitions for the fields of packets,
https://github.com/cosmos/ibc/blob/jk/ccvTLA/spec/app/ics-028-cross-chain-validation/data_structures.md#ccv-state for the ccv timeout

Does that make sense? Do you think there's anything I could add to the code to make it easier to understand?